src/metrics/tagfiltertree/tag_filter_tree.go

```diff
@@ -0,0 +1,317 @@
+	Tree        *Tree[T]
+	Data        []T
+}
+
+type node[T any] struct {
+	Name           string
+	AbsoluteValues map[string]NodeValue[T]
+	PatternValues  []NodeValue[T]
+}
+
+func (n *node[T]) addValue(filter string, data *T) (*Tree[T], error) {
+	if IsAbsoluteValue(filter) {
+		if n.AbsoluteValues == nil {
+			n.AbsoluteValues = make(map[string]NodeValue[T], 0)
+		}
+		if v, found := n.AbsoluteValues[filter]; found {
+			if data != nil {
+				if v.Data == nil {
+					v.Data = make([]T, 0)
+				}
+				v.Data = append(v.Data, *data)
+				n.AbsoluteValues[filter] = v
+			}
+			return v.Tree, nil
+		}
+
+		newNodeValue := NewNodeValue[T](filter, nil, data)
+		n.AbsoluteValues[filter] = newNodeValue
+
+		return newNodeValue.Tree, nil
+	}
+
+	if n.PatternValues == nil {
+		n.PatternValues = make([]NodeValue[T], 0)
+	}
+	for i, v := range n.PatternValues {
+		if v.Val == filter {
+			if data != nil {
+				if v.Data == nil {
+					v.Data = make([]T, 0)
+				}
+				v.Data = append(v.Data, *data)
+				n.PatternValues[i] = v
+			}
+			return v.Tree, nil
+		}
+	}
+
+	t := NewTrie[T]()
+	err := t.Insert(filter, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	newNodeValue := NewNodeValue[T](filter, t, data)
+	n.PatternValues = append(n.PatternValues, newNodeValue)
+
+	return newNodeValue.Tree, nil
+}
+
+// NewNodeValue creates a new NodeValue.
+func NewNodeValue[T any](val string, patternTrie *Trie[T], data *T) NodeValue[T] {
+	t := &Tree[T]{
+		Nodes: make([]*node[T], 0),
+	}
+
+	v := NodeValue[T]{
+		Val:         val,
+		PatternTrie: patternTrie,
+		Tree:        t,
+		Data:        nil,
+	}
+
+	if data != nil {
+		if v.Data == nil {
+			v.Data = make([]T, 0)
+		}
+		v.Data = append(v.Data, *data)
+	}
+
+	return v
+}
+
+func addNode[T any](t *Tree[T], tags []Tag, idx int, data T) error {
+	if idx >= len(tags) {
+		return nil
+	}
+
+	tag := tags[idx]
+	nodeIdx := -1
+	for i, t := range t.Nodes {
+		if t.Name == tag.Name {
+			nodeIdx = i
+			break
+		}
+	}
+	if nodeIdx == -1 {
+		t.Nodes = append(t.Nodes, &node[T]{
+			Name: tag.Name,
+		})
+		nodeIdx = len(t.Nodes) - 1
+	}
+	node := t.Nodes[nodeIdx]
+
+	// Add the data to the childTree if this is the last tag.
+	var dataToAdd *T
+	if idx == len(tags)-1 {
+		dataToAdd = &data
+	}
+
+	childTree, err := node.addValue(tag.Val, dataToAdd)
+	if err != nil {
+		return err
+	}
+
+	// Recurse to the next tag.
+	if err := addNode(childTree, tags, idx+1, data); err != nil {
+		// TODO: perform cleanup to avoid partially added nodes.
+		return err
+	}
+
+	return nil
+}
+
+func match[T any](
+	t *Tree[T],
+	tags map[string]string,
+	data *[]T,
+	isMatchAny bool,
+) (bool, error) {
+	if len(tags) == 0 || t == nil {
+		return false, nil
+	}
+
+	for _, node := range t.Nodes {
+		name := node.Name
+		negate := false
+		if IsMatchNoneTag(name) {
+			name = name[1:]
+			negate = true
+		}
+		tagValue, tagNameFound := tags[name]
+		absVal, absValFound := node.AbsoluteValues[tagValue]
+		if tagNameFound && absValFound {
```

Comment:
if isMatchAny==false, then within this if condition, you will never return true? is that correct logic?

Reply:
Ha! good catch! yeah this is a bug. We cannot rely on the returned "matched" bool flag. I'll fix it.